jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/HttpResponseStatisticsCollector.java

```diff
@@ -0,0 +1,199 @@
+        }
+
+        @Override
+        public void onError(AsyncEvent event) throws IOException {
+        }
+
+        @Override
+        public void onComplete(AsyncEvent event) throws IOException {
+            HttpChannelState state = ((AsyncContextEvent) event).getHttpChannelState();
+            Request request = state.getBaseRequest();
+
+            observeEndOfRequest(request, null);
+        }
+    };
+
+    @Override
+    public void handle(String path, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
+            throws IOException, ServletException {
+        inFlight.incrementAndGet();
+
+        try {
+            Handler handler = getHandler();
+            if (handler != null && shutdown.get() == null && isStarted()) {
+                handler.handle(path, baseRequest, request, response);
+            } else if (!baseRequest.isHandled()) {
+                baseRequest.setHandled(true);
+                response.sendError(HttpStatus.SERVICE_UNAVAILABLE_503);
+            }
+        } finally {
+            HttpChannelState state = baseRequest.getHttpChannelState();
+
+            if (state.isSuspended()) {
+                if (state.isInitial()) {
+                    state.addListener(completionWatcher);
+                }
+            } else if (state.isInitial()) {
+                observeEndOfRequest(baseRequest, response);
+            }
+        }
+    }
+
+    private void observeEndOfRequest(Request request, HttpServletResponse flushableResponse) throws IOException {
+        int group = groupIndex(request);
+        if (group >= 0) {
+            HttpMethod method = getMethod(request);
+            statistics[method.ordinal()][group].increment();
+        }
+
+        long live = inFlight.decrementAndGet();
+        FutureCallback shutdownCb = shutdown.get();
+        if (shutdownCb != null) {
+            if (flushableResponse != null) {
+                flushableResponse.flushBuffer();
+            }
+            if (live == 0) {
+                shutdownCb.succeeded();
+            }
+        }
+    }
+
+    private int groupIndex(Request request) {
+        if (request.isHandled()) {
+            int index = (request.getResponse().getStatus() / 100) - 1; // 1xx = 0, 2xx = 1 etc.
+            if (index < 0 || index > statistics.length) {
+                return -1;
+            } else {
+                return index;
+            }
+        } else {
+            return 3; // 4xx
+        }
+    }
+
+    private HttpMethod getMethod(Request request) {
+        switch (request.getMethod()) {
+        case "GET":
+            return HttpMethod.GET;
+        case "PATCH":
+            return HttpMethod.PATCH;
+        case "POST":
+            return HttpMethod.POST;
+        case "PUT":
+            return HttpMethod.PUT;
+        case "DELETE":
+            return HttpMethod.REMOVE;
+        default:
+            return HttpMethod.UNKNOWN;
```

Comment:
OPTIONS and HEAD are now reported as _unknown_. Why are they not separate enum values?

Reply:
Oversight based on the request for the feature. I added those and renamed `UNKNOWN` to `OTHER` to better reflect the status.